sycl/doc/design/SYCL-RTC.md

```diff
@@ -0,0 +1,312 @@
+well as from the filesystem, e.g. `sycl.hpp` from the DPC++ installation or user
+libraries, it is not sufficient to look only at the source string. In order to
+make the cache as conservative as possible (cache collisions are unlikely but
+mathematically possible), we decided to compute a hash value of the
+*preprocessed* source string, i.e. with all `#include` directives resolved. We
+additionally compute a hash value of the rendered command-line arguments, and
+append it to the hash of the preprocessed source to obtain the final cache key. 
+
+### Implementation notes
+
+The cache key computation is implemented in the
+[`jit_compiler::calculateHash(...)`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp#L381)
+function. We are again relying on LibTooling to invoke the preprocessor -
+handily, Clang provides  a
+[`PreprocessorFrontendAction`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/clang/include/clang/Frontend/FrontendAction.h#L294)
+that we extend to tailor to our use-case. We choose
+[BLAKE3](https://en.wikipedia.org/wiki/BLAKE_(hash_function)) as the hash
+algorithm because its proven in similar contexts (most notably,
+[ccache](https://ccache.dev)) and available as a utility in the LLVM ecosystem.
+As the output is a byte array, we apply Base64 encoding to obtain a character
+string for use with the persistent cache.
+
+## Device library linking and SYCL-specific transformations
+
+With an LLVM IR module in hand, obtained either from the frontend or the cache,
+the next steps in the compilation pipeline are simple.
+
+The device library linking is done by the
+[`jit_compiler::linkDeviceLibraries(...)`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp#L566)
+function. These libraries provide primitives for a variety of extra
+functionality, such as an extended set of math functions and support for
+`bfloat16` arithmetic, and are available as Bitcode files inside the DPC++
+installation or the vendor toolchain, so we just use LLVM utilities to load them
+into memory and link them to the module representing the runtime-compiled
+kernels.
+
+For the SYCL-specific post-processing, implemented in
+[`jit_compiler::performPostLink(...)`](https://github.com/intel/llvm/blob/cc966df07d29db75d07f969f044c0491819bd930/sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp#L750),
+we can reuse modular analysis and transformation passes in the
+[`SYCLLowerIR`](https://github.com/intel/llvm/tree/sycl/llvm/lib/SYCLLowerIR)
+component. The main tasks for the post-processing passes is to split the device
+code module into smaller units (either as requested by the user, or required by
+the ESIMD mode), and to compute the properties that need to be passed to the
+SYCL runtime when the device images are loaded.
+
```

Comment:
We could mention the code duplication and the plans to remove it in the future here.